### ISSUE TITLE
BasicVSR++2 and Training Speed

### ISSUE THREAD

**Issue by mikolez** [ckkelvinchan/BasicVSR_PlusPlus#21]
In the technical report for deblurring and denoising (https://arxiv.org/pdf/2204.05308.pdf), there are two versions of the BasicVSR++ for video denoising, the one which reduces the resolution by 2 (BasicVSR++2, superior performance but less speed) and by 4 (BasicVSR++4, more speed but worse performance). Current code, as I understand, implements the version which reduces the resolution by 4. How do I change the code exactly to switch to the BasicVSR++2 (the one which reduces the resolution by 2)? Or could you please share the code?

In addition, I am trying to speed up the training for denoising. I tried increasing samples_per_gpu config variable from 1 to 3 and decreasing the number of iterations from 600,000 to 200,000 (so that the amount of data that the model uses for training is the same), which helped to improve the time quite well (from 13 days to 6 days of training). Is there any other way to make the training faster without hurting the performance after the training?

Thank you!

**Comment by Dylan-Jinx** [comment]
Hello,I have a similar problem.  Could you please tell me you choose dataset, is REDS or DIV2K?

**Comment by ckkelvinchan** [comment]
Hello @mikolez and @Dylan-Jinx, sorry for the (very) late reply. 

@mikolez 
1. To reduce the resolution by 2 times instead of 4, you can just simply reduce the number of strided conv and pixel shuffle from 2 to 1. I will share the code and model soon.
2. About the training time, you may try to decrease the sequence length and further increase the batch size. I am not sure whether the performance will be better or worse. About increasing the batch size, do you obtain the similar performance?

@Dylan-Jinx 
DIV2K is an image dataset. We use `REDS` for SR, `DVD/GoPro` for Deblur, and `DAVIS` for Denoise.

**Comment by AIisCool** [comment]
@ckkelvinchan look forward to seeing the improvement BasicVSR++2 can have!

**Comment by chaojihongjing** [comment]
> Hello @mikolez and @Dylan-Jinx, sorry for the (very) late reply.
> 
> @mikolez
> 
> 1. To reduce the resolution by 2 times instead of 4, you can just simply reduce the number of strided conv and pixel shuffle from 2 to 1. I will share the code and model soon.
> 2. About the training time, you may try to decrease the sequence length and further increase the batch size. I am not sure whether the performance will be better or worse. About increasing the batch size, do you obtain the similar performance?
> 
> @Dylan-Jinx
> 
> 1. DIV2K is an image dataset. We use `REDS` for SR, `DVD/GoPro` for Deblur, and `DAVIS` for Denoise.

@ckkelvinchan Hi, thank you very much for your excellent work. Could you please tell me the dataset for Decompress training?

**Comment by ckkelvinchan** [comment]
> @ckkelvinchan look forward to seeing the improvement BasicVSR++2 can have!

Hi @AIisCool, I have uploaded the checkpoints for BasicVSR_2x. You can find the checkpoints in the same Dropbox link mentioned in README. I will try to upload the code this week, but it should be pretty trivial to modify as you only need to remove one strided conv and one pixel-shuffling.

Sorry to keep you waiting, enjoy!

**Comment by AIisCool** [comment]
> I have uploaded the checkpoints for BasicVSR_2x

Excellent thank you! I will have to wait for the code as I am not too sure what to modify exactly.

**Comment by bgalerne** [comment]
Hi,

First thanks for your work and for sharing the code. 

Is the code for using the released 

[basicvsr_plusplus_2x_denoise.pth
](https://www.dropbox.com/sh/gopo637rbutlxde/AABccNWk5LHRiZ6x7vFedbTsa/basicvsr_plusplus_2x_denoise.pth)

available yet ? Thank you in advance.

**Comment by zhu2bowen** [comment]
@AIisCool Do you know how to fix the code for x2 super resolutiuon now?

**Comment by ioctl-user** [comment]
>     1. To reduce the resolution by 2 times instead of 4, you can just simply reduce the number of strided conv and pixel shuffle from 2 to 1. I will share the code and model soon.

So, what changes should be done in the https://github.com/ckkelvinchan/BasicVSR_PlusPlus/blob/master/configs/basicvsr_plusplus_denoise.py to use 2-times resolution model?

**Comment by lcc157** [comment]
> > ```
> > 1. To reduce the resolution by 2 times instead of 4, you can just simply reduce the number of strided conv and pixel shuffle from 2 to 1. I will share the code and model soon.
> > ```
> 
> So, what changes should be done in the https://github.com/ckkelvinchan/BasicVSR_PlusPlus/blob/master/configs/basicvsr_plusplus_denoise.py to use 2-times resolution model?

Do you know how to fix the code for x2 super resolutiuon now?

**Comment by ioctl-user** [comment]
> > > ```
> > > 1. To reduce the resolution by 2 times instead of 4, you can just simply reduce the number of strided conv and pixel shuffle from 2 to 1. I will share the code and model soon.
> > > ```
> > 
> > 
> > So, what changes should be done in the https://github.com/ckkelvinchan/BasicVSR_PlusPlus/blob/master/configs/basicvsr_plusplus_denoise.py to use 2-times resolution model?
> 
> Do you know how to fix the code for x2 super resolutiuon now?

Still not.

**Comment by PaatrickStaar** [comment]
> Hello @mikolez and @Dylan-Jinx, sorry for the (very) late reply.
> 
> @mikolez
> 
> 1. To reduce the resolution by 2 times instead of 4, you can just simply reduce the number of strided conv and pixel shuffle from 2 to 1. I will share the code and model soon.
> 2. About the training time, you may try to decrease the sequence length and further increase the batch size. I am not sure whether the performance will be better or worse. About increasing the batch size, do you obtain the similar performance?
> 
> @Dylan-Jinx
> 
> 1. DIV2K is an image dataset. We use `REDS` for SR, `DVD/GoPro` for Deblur, and `DAVIS` for Denoise.

I can reduce the training time by decreasing the frame rate to 15 frames, but when I modify samples_per_gpu (which is equivalent to batch size) in train_dataloader=dict(samples_per_gpu=1, drop_last=True) in the config file, it actually increases the training time. Why is this the case?

**Comment by sondv2** [comment]
Hi @ckkelvinchan 
I got this error 

python demo/restoration_video_demo.py configs/basicvsr_plusplus_deblur_dvd.py chkpts/basicvsr_plusplus_2x_deblur_dvd.pth data/tayduky_360p results/tayduky_360p_2x_deblur_dvd

2024-08-16 10:04:58,025 - mmedit - INFO - load checkpoint from http path: https://download.openmmlab.com/mmediting/restorers/basicvsr/spynet_20210409-c6c1bd09.pth
load checkpoint from local path: chkpts/basicvsr_plusplus_2x_deblur_dvd.pth
The model and loaded state dict do not match exactly

unexpected key in source state_dict: generator.feat_extract.2.main.0.weight, generator.feat_extract.2.main.0.bias, generator.feat_extract.2.main.2.0.conv1.weight, generator.feat_extract.2.main.2.0.conv1.bias, generator.feat_extract.2.main.2.0.conv2.weight, generator.feat_extract.2.main.2.0.conv2.bias, generator.feat_extract.2.main.2.1.conv1.weight, generator.feat_extract.2.main.2.1.conv1.bias, generator.feat_extract.2.main.2.1.conv2.weight, generator.feat_extract.2.main.2.1.conv2.bias, generator.feat_extract.2.main.2.2.conv1.weight, generator.feat_extract.2.main.2.2.conv1.bias, generator.feat_extract.2.main.2.2.conv2.weight, generator.feat_extract.2.main.2.2.conv2.bias, generator.feat_extract.2.main.2.3.conv1.weight, generator.feat_extract.2.main.2.3.conv1.bias, generator.feat_extract.2.main.2.3.conv2.weight, generator.feat_extract.2.main.2.3.conv2.bias, generator.feat_extract.2.main.2.4.conv1.weight, generator.feat_extract.2.main.2.4.conv1.bias, generator.feat_extract.2.main.2.4.conv2.weight, generator.feat_extract.2.main.2.4.conv2.bias

missing keys in source state_dict: generator.feat_extract.2.weight, generator.feat_extract.2.bias, generator.feat_extract.4.main.0.weight, generator.feat_extract.4.main.0.bias, generator.feat_extract.4.main.2.0.conv1.weight, generator.feat_extract.4.main.2.0.conv1.bias, generator.feat_extract.4.main.2.0.conv2.weight, generator.feat_extract.4.main.2.0.conv2.bias, generator.feat_extract.4.main.2.1.conv1.weight, generator.feat_extract.4.main.2.1.conv1.bias, generator.feat_extract.4.main.2.1.conv2.weight, generator.feat_extract.4.main.2.1.conv2.bias, generator.feat_extract.4.main.2.2.conv1.weight, generator.feat_extract.4.main.2.2.conv1.bias, generator.feat_extract.4.main.2.2.conv2.weight, generator.feat_extract.4.main.2.2.conv2.bias, generator.feat_extract.4.main.2.3.conv1.weight, generator.feat_extract.4.main.2.3.conv1.bias, generator.feat_extract.4.main.2.3.conv2.weight, generator.feat_extract.4.main.2.3.conv2.bias, generator.feat_extract.4.main.2.4.conv1.weight, generator.feat_extract.4.main.2.4.conv1.bias, generator.feat_extract.4.main.2.4.conv2.weight, generator.feat_extract.4.main.2.4.conv2.bias, generator.upsample2.upsample_conv.weight, generator.upsample2.upsample_conv.bias

**Comment by kevinrex66** [comment]
/mmedit/models/backbones/sr_backbones/basicvsr_pp.py
line 308 
        #lqs.view(-1, c, h, w), scale_factor=0.25,
        lqs.view(-1, c, h, w), scale_factor=0.5,
line 269
        #hr = self.lrelu(self.upsample2(hr))
line 68:
        #         nn.Conv2d(mid_channels, mid_channels, 3, 2, 1),
        #         nn.LeakyReLU(negative_slope=0.1, inplace=True),
line 87:
        # # upsampling module
        # self.reconstruction = ResidualBlocksWithInputConv(
        #     5 * mid_channels, mid_channels, 5)
        # self.upsample1 = PixelShufflePack(
        #     mid_channels, mid_channels, 2, upsample_kernel=3)
        # self.upsample2 = PixelShufflePack(
        #     mid_channels, 64, 2, upsample_kernel=3)
        # self.conv_hr = nn.Conv2d(64, 64, 3, 1, 1)
        # self.conv_last = nn.Conv2d(64, 3, 3, 1, 1)
        # self.img_upsample = nn.Upsample(
        #     scale_factor=4, mode='bilinear', align_corners=False)
        # upsampling module (one ×2 pixel shuffle + 2× bilinear)
        self.reconstruction = ResidualBlocksWithInputConv(
            5 * mid_channels, mid_channels, 5)
        self.upsample1 = PixelShufflePack(
            mid_channels, 64, 2, upsample_kernel=3)
        self.conv_hr = nn.Conv2d(64, 64, 3, 1, 1)
        self.conv_last = nn.Conv2d(64, 3, 3, 1, 1)
        self.img_upsample = nn.Upsample(
            scale_factor=2, mode='bilinear', align_corners=False)
